v2.1/bytes.md

```diff
@@ -66,6 +66,33 @@ valid UTF-8 byte sequences; an error is reported otherwise.
 `STRING` values can be cast explicitly to `BYTES`. This conversion
 always succeeds.
 
+### `STRING` vs. `BYTES`
+
+While both `STRING` and `BYTES` can appear to have similar behavior in many situations, one should understand their nuance before casting one into the other.
+
+`STRING` treats all of its data as characters, or more specificially, unicode code points. `BYTES` treats all of its data as a byte string. This difference in implementation can lead to dramatically different behavior. For example:
+
+{% include copy-clipboard.html %}
+~~~ sql
+CREATE TABLE f (x STRING, y BYTES);
+~~~
+~~~
+CREATE TABLE
+~~~
+~~~ sql
+INSERT INTO f (x, y) VALUES ('\xff', b'\xff'); 
+SELECT LENGTH(x), LENGTH(y) FROM f;
+~~~
+~~~
++-----------+-----------+
+| length(x) | length(y) |
++-----------+-----------+
+|         4 |         1 |
++-----------+-----------+
+~~~
+
+In this case, `LENGTH(bytes)` measures the number of bytes; whereas `LENGTH(string)` measures the number of unicode code points present in the string. Each character (or unicode code point) can be encoded using multiple bytes, hence the difference in output between the two.
```

Comment:
This is conflating multiple steps together unfortunately, to the point it becomes (slightly) inaccurate.

Maybe start with this example:

```
root@127.1.169.208:44415/defaultdb> select length('☃'::string);
  length
+--------+
       1
(1 row)


root@127.1.169.208:44415/defaultdb> select length('☃'::bytes);
  length
+--------+
       3
(1 row)

```

This will support your argument above.

Then *after* you make that argument, you should also talk about the process of translating a *literal* (what the client enters in the SQL syntax) to a STRING or BYTES *value* inside CockroachDB.

There, again different rules apply:

- BYTES will give a special meaning to the pair `\x` at the beginning, and translate the rest by substituting pairs of hexadecimal digits to a single byte. This is why `\xff` gives just 1 byte (with value 255).
- STRING does not give a special meaning to `\x` and just keeps every character (including `\` and `x`) in the result. This is why `\xff` gives 4 characters (`\`, `x`, `f`, `f`).